### PR TITLE
Add error section

### PR DESCRIPTION
**Pull request by duckthecuck**
This PR adds a section on how to diagnose errors, different types of errors, and some common errors, it is HEAVILY experimental, all feedback is appreciated, i am not quite sure how to structure this, so if you have ideas please comment them. ~~For right now, websocket errors are not included, but they could possibly be added to fulfil #195~~ Websocket errors are partially added. This section is planned to fulfil https://github.com/discordjs/guide/projects/1#card-16181091

### CODE REVIEW

**Comment by duckthecuck** [comment]
If you have any errors you would like to see added just comment them, i cant remember all common errors 👀

**Comment by duckthecuck** [comment]
With regards to the Websocket and network errors, some are thrown by the websocket, and some arent but are still network errors, i am unsure whether to include the others like `ECONNRESET` and `ENOTFOUND` .

File: guide/popular-topics/errors.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+
+### JS Errors
+
+JS Errors are simple errors which can be thrown by node itself, or by discord.js. These types of errors can easily be fixed by looking at the type of error, and the stacktrace. You can find a full list of types [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) And a list of common js errors [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors).
+
+Example: `ReferenceError: "x" is not defined`
+
+### Websocket and Network Errors
+
+Websocket and Network errors are common system errors thrown by Node in response to something wrong with the websocket connection. Unfortunately, these errors do not have a concrete solution and can be (usually) fixed by getting a better, and stronger connection. However, if you handle these errors correctly, discord.js will attempt to reconnect to the websocket automatically.
```
Comment (vladfrangu) [inline guide/popular-topics/errors.md]: This isn't really accurate, definitely not on master. We are handling network errors internally and attempt to reconnect if you have a network blip, in order to reconnect you the moment we can. 

> Unfortunately, these errors do not have a concrete solution and can be (usually) fixed by getting a better, and stronger connection

More-less; you need a stable connection too!

> However, if you handle these errors correctly, discord.js will attempt to reconnect to the websocket automatically.

In master, we emit `shardError` for any WS error we receive from the WS, the rest being handled internally to ensure you, the end user, have no issues even if your internet randomly dies. I don't know how applicable this is to stable d.js so don't quote me

File: guide/popular-topics/errors.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+You can handle these errors by adding an error listener as shown below.
```
Comment (vladfrangu) [inline guide/popular-topics/errors.md]: "Handle" is a bit of a *stretch* — you're preventing the process from crashing, sure, but you aren't really... doing anything with the error; just logging it. Maybe a slight reword would work

File: guide/popular-topics/errors.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+```js
+client.on('error', console.error);
+```
+
+Now, when an error occurs it will be logged to the console and discord.js will automatically attempt to reconnect to the websocket.
```
Comment (vladfrangu) [inline guide/popular-topics/errors.md]: As said above, at least in master, you don't need to log them, d.js will still attempt to reconnect you

File: guide/popular-topics/errors.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+</branch>
+
+### The messages must be an Array, Collection, or number.
+
+This error originates from an invalid call to `bulkDelete()`, make sure you are inputting a valid Array or Collection of messages, or a valid number.
+
+### Members didnt't arrive in time.
+
+Another common error, this error originates from the client requesting members from the API through the websocket, and the member chunks not arriving in time and triggering the timeout. The most common cause to this error is a bad connection, however, it can also be caused by a very large amount of members being fetched, upwards of 50 thousand. To fix this, run the bot on a location with better internet, such as a VPS. If this does not work for you, you will have to manually change the member fetching timeout.
```
Comment (vladfrangu) [inline guide/popular-topics/errors.md]: Just fyi: The timeout, at least in master, is hard coded in. Editing would mean patching one line from internals; definitely not a hard task, but for beginners it sure isn't the simplest too.

File: guide/popular-topics/errors.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+- The user has blocked the bot.
+- The user has disabled dms in the privacy settings.
+
+In the case of the last two reasons, the error is not preventable, as the Discord API does not provide a way to check if you can send a user a dm until you attempt to send one. The best way to handle this error is to add a `.catch()` where you attempt to dm the user, and either ignore the rejected promise, or do what you want because of it.
+
+## Common miscellaneous errors
+
+### code ENOENT... syscall spawn git.
+
+This is an error commonly thrown by npm due to git not being installed on your machine. You must install git:
```
Comment (vladfrangu) [inline guide/popular-topics/errors.md]: *Technically* this error is emitted by system -> node.js -> npm/yarn. Maybe reword this a bit?

File: guide/popular-topics/errors.md (hunk 1 above)
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: i know there are a lot of websocket changes between stable and master, so i was not very sure if master already handled reconnecting without the error listener. For master should i mention you should listen to `shardError` to be notified about websocket errors? and that the lib will already attempt to reconnect?

File: guide/popular-topics/errors.md (hunk 2 above)
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: i agree, i will reword it to something along the lines of `you can prevent these errors from crashing your process with`..., would that be good?

File: guide/popular-topics/errors.md (hunk 3 above)
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: yep, will refactor all of this for master

File: guide/popular-topics/errors.md (hunk 4 above)
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: yeah, i know about that, i forgot to mention its hardcoded :^), one of the things i dont really get, in my opinion it should be able to be easily edited through client options or somewhere else, but thats another topic to discuss.

File: guide/popular-topics/errors.md (hunk 1 above)
Comment (vladfrangu) [reply guide/popular-topics/errors.md]: Both versions handle reconnects automatically, the issue is that stable emits the `error` event, which by default throws if not handled (thus, crashing). 

You don't *have* to listen to `shardError` at all! The lib should work regardless of it

Feel free to amend it like the the MessageField changes if you're reading the v12.x branch, if you want

File: guide/popular-topics/errors.md (hunk 5 above)
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: `This is an error commonly thrown by Node in response to npm not being able to find git`...?

File: guide/popular-topics/errors.md (hunk 2 above)
Comment (vladfrangu) [reply guide/popular-topics/errors.md]: Yep yep, sounds good! Then aaagain, console.log is just a fancier no-op to prevent the process from crashing, you should handle those errors regardless

File: guide/popular-topics/errors.md (hunk 1 above)
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: yeah i think ill refactor master websocket stuff to say 
`Master will automatically handle websocket connection errors and attempt to reconnect, if you would like to do something based on these errors, you should listen to shardError`, does that look good to you?

File: guide/popular-topics/errors.md (hunk 4 above)
Comment (vladfrangu) [reply guide/popular-topics/errors.md]: Feel free to make it a suggestion on the main repo if you want!

File: guide/popular-topics/errors.md (hunk 2 above)
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: yeah, its just an issue of "the only real solution is get a better connection" to handle these errors

File: guide/popular-topics/errors.md (hunk 1 above)
Comment (vladfrangu) [reply guide/popular-topics/errors.md]: ``On master, WebSocket errors are handled internally, meaning your process should never crash from them. If you want to log these errors, should they happen, you can listen to the `shardError` event``

File: guide/popular-topics/errors.md (hunk 4 above)
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: yeah, i feel like this will be helpful for people with a large amount of guilds, i will submit a PR/issue to add this later, when i have some time

File: guide/popular-topics/errors.md (hunk 1 above)
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: sounds good

File: guide/popular-topics/errors.md (hunk 5 above)
Comment (vladfrangu) [reply guide/popular-topics/errors.md]: ``This error is commonly thrown by your system due to it not being able to find `git`. You need to install `git` or update your path if `git` is already installed. Here are the download links for it:
- Links 
- Here
```
Or something that doesn't sound bad 😄

File: guide/popular-topics/errors.md (hunk 4 above)
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: do you happen to know where this interval is hardcoded?

Comment (vladfrangu) [reply guide/popular-topics/errors.md]: This is where the timeout is created

https://github.com/discordjs/discord.js/blob/50ed3293a57aeb0f9c094f00d49ed9874dc5d87a/src/stores/GuildMemberStore.js#L221-L224

I recommend creating a issue truthfully, since a better approach would be refreshing the timeout with every packet we get; but feel free to submit a PR too!

Comment (duckthecuck) [reply guide/popular-topics/errors.md]: Alright, thank you

**Comment by duckthecuck** [comment]
Websocket is a hell of a drug, let me know if this looks good @vladfrangu

**Comment by vladfrangu** [comment]
> Websocket is a hell of a drug, let me know if this looks good @vladfrangu

You tell me! But yeah, this looks alright to me!

**Comment by duckthecuck** [comment]
Ill go ahead and remove the [WIP] since it is mostly finished unless someone wants other errors to be added or wants something changes

File: guide/popular-topics/errors.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+
+API Errors or DiscordAPIErrors are errors which are thrown by the Discord API when an invalid request is carried out. API Errors can be mostly diagnosed using the message that is given. They can also be further examined by seeing the http method and path used, we will explore tracking these errors down in the next section.
+
+Example: `DiscordAPIError: Cannot send an empty message`
+
+### Discord.js Errors
+
+Discord.js Errors are errors which are thrown by the library itself, they can usually be easily tracked down using the stacktrace and error message.
+
+Example: `Error: Members didn't arrive in time.`
```
Comment (SpaceEEC) [inline guide/popular-topics/errors.md]: Just as a note: This error is not easily tracked down since the stacktrace will only point to timers and the library itself, not user code.
This example should probably use a different error (maybe one from RichEmbed/MessageEmbed or failing to resolve role to add / remove, something *sync*).

File: guide/popular-topics/errors.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+In this particular example we can see we are trying to access a piece of data, specifically, a message.
+
+## Common Discord.js and API errors
+
+### An invalid token was provided.
+
+This is a very common error, it originates from a wrong token being passed into `client.login()`. The most common causes of this error are:
+
+- Not importing the config or env file correctly
+- Copying the client secret instead of the bot token (token is alphanumerical and 3 parts delimited by a period while client secret is all numbers)
```
[the remark this answers is not in the document]
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: Oops, remembered the client id, you're right

File: guide/popular-topics/errors.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+The most important part of the error is the message, it tells you what actually went wrong which can help you track down where it came from. 
+You can find a full list of messages [here](https://discordapp.com/developers/docs/topics/opcodes-and-status-codes#json) in the Discord API Docs.
+
+### Path
+
+The path is another helpful piece of information, the path tells you where you tried to execute an action. We cannot possibly cover all API paths but they are usually very descriptive, for example, in our example above the path tells us we are first in the channels path, then judging by the id after it we can see we got a specific channel. After the `/api/v7/channels/638200642359525387` we can see that we get the `messages` of that channel, and in the same way we saw before, we see we try to access a specific message by id, however we gave it an invalid id, which is the origin of the message.
+
+### Code
+
+The code is another partial representation of the message, in this case `Invalid Form Body`. You can find a full list of codes [here](https://discordapp.com/developers/docs/topics/opcodes-and-status-codes#json-json-error-codes)
```
[the remark this answers is not in the document]
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: what do you mean by handle by code? just check if code is something and do something? in that case, what would be the something? im not sure i see a use

[the remark this answers is not in the document]
Comment (duckthecuck) [reply guide/popular-topics/errors.md]: maybe something like 
> If you would like to handle specific errors, you can simply check the code. You can copy the code from the discord API Docs or you can use the discord.js constants.

> Example:
```js
process.on('unhandledRejection', error => {
	if(error.code === 10008) {
            console.error('Unknown message error: ', error);
        }
});```

**Comment by duckthecuck** [comment]
@SpaceEEC the example added showcases ignoring a specific error, not handling a specific error, but i feel that is the most common use case, and it still showcases how to use the code

File: guide/popular-topics/errors.md [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+- Through the [CLI](https://nodejs.org/api/cli.html#cli_trace_warnings): `node --trace-warnings index.js`
+- Through the [`process#warning` event](https://nodejs.org/api/process.html#process_event_warning): `process.on('warning', console.warn);`
+
+### Cannot send messages to this user.
+
+This error is thrown when the bot attempts to send a DM message to a user and it is unable to do so. This is caused by a variety of reasons:
+- The bot and the user do not share a guild (oftentimes people accidentally kick or ban a user and then try to dm them).
+- The user has blocked the bot.
+- The user has disabled dms in the privacy settings.
+- The bot is attempting to DM another bot.
```
Comment (SpaceEEC) [inline guide/popular-topics/errors.md]: You should move this up since the below text says `In the case of the last two reasons`.

Comment (duckthecuck) [reply guide/popular-topics/errors.md]: woops
